File: ds_and_algo/implementations/python/algorithms/searching/linear_search.py
# Arrays (Lists in Python):

array = ["a", False, 5]

# array.indexOf("a")
print(array.index("a"))  # 0
# array.findIndex((item) => typeof item === 'boolean')
print(next(i for i, item in enumerate(array) if isinstance(item, bool)))  # 1
# array.find((item) => item > 3)
print(next(item for item in array if isinstance(item, int) and item > 3))  # 5
# array.includes("a")
print("a" in array)  # True

# Hash Tables (Dictionaries in Python):
people = {
    1: {"name": "Alice", "age": 20},
    2: {"name": "Bob", "age": 25},
    3: {"name": "Charlie", "age": 30},
}

# Find the person whose age is greater than 25:
found_person = next((id, person) for id, person in people.items() if person["age"] > 25)
print(found_person)  # (3, {'name': 'Charlie', 'age': 30})

# Find the ID of the person whose name is Bob:
bob_id = next(key for key, value in people.items() if value["name"] == "Bob")
print(bob_id)  # 2

# Check for a given name:
name_exists = any(person["name"] == "Bob" for person in people.values())
print(name_exists)  # True

File: ds_and_algo/implementations/javascript/algorithms/searching/linear-search.js
// Arrays

const array = ["a", false, 5];

console.log(array.indexOf("a")); // 0
console.log(array.findIndex((item) => typeof item === "boolean")); // 1
console.log(array.find((item) => item > 3)); // 5
console.log(array.includes("a")); // true

// Hash Tables

/*
General Recipe:

1. Convert the Hash Table into an Array.
2. Apply built-in Array methods.

- To use `.find()` and `.findIndex()`, convert the object into an Array of entries using `Object.entries()`.

- To use `.indexOf` and `.includes()`, convert the object to an Array of values using `Object.values()`.
*/

const people = {
  1: { name: "Alice", age: 20 },
  2: { name: "Bob", age: 25 },
  3: { name: "Charlie", age: 30 },
};

// Find the person whose age is greater than 25:
let foundPerson = Object.entries(people).find(([, person]) => person.age > 25);
console.log(foundPerson); // [ '3', { name: 'Charlie', age: 30 } ]

// Find the ID of the person whose name is Bob:
let foundIndex = Object.keys(people).find((key) => people[key].name === "Bob");
console.log(foundIndex); // 2

// Check for a given name:
let nameExists = Object.values(people)
  .map((person) => person.name)
  .includes("Bob");

console.log(nameExists); // true
